kani-compiler/src/codegen_cprover_gotoc/codegen/operand.rs

```diff
@@ -178,279 +268,69 @@ impl<'tcx> GotocCtx<'tcx> {
             }
-            (Scalar::Int(int), ty::Tuple(_)) => {
-                // A ScalarInt has a u128-typed data field, so the result can never be larger than
-                // that and the conversion to a uint (of an actual size that may be smaller than
-                // 128 bits) will succeed.
-                let int_u128 = int.try_to_uint(int.size()).ok().unwrap();
-                let overall_t = self.codegen_ty(ty);
-                let expr_int = Expr::int_constant(
-                    int_u128,
-                    Type::unsigned_int(overall_t.sizeof_in_bits(&self.symbol_table)),
-                );
-                expr_int.transmute_to(overall_t, &self.symbol_table)
-            }
-            (_, ty::Array(_, _)) => {
-                let typ = self.codegen_ty(ty);
-                // we must have zero size array here
-                Expr::struct_expr_from_values(
-                    typ.clone(),
-                    vec![Expr::array_expr(typ, vec![])],
-                    &self.symbol_table,
-                )
-            }
-            (Scalar::Ptr(ptr, _size), _) => {
-                let res_t = self.codegen_ty(ty);
-                let (prov, offset) = ptr.into_parts();
-                let alloc_id = prov.alloc_id();
-                self.codegen_alloc_pointer(res_t, alloc_id, offset, span)
-            }
-            _ => unimplemented!(),
-        }
-    }
-
-    /// A private helper for `codegen_scalar`. Many "scalars" are more complex types, but get treated as scalars
-    /// because they only have one (small) field. We still translated them as struct types, however.
-    fn codegen_single_variant_single_field(
-        &mut self,
-        s: Scalar,
-        span: Option<&Span>,
-        overall_t: Type,
-        fty: Ty<'tcx>,
-    ) -> Expr {
-        if fty.is_unit() {
-            // TODO: It's not clear if this case is reachable. It's not covered by our test suite at least.
-            Expr::struct_expr_from_values(overall_t, vec![], &self.symbol_table)
+        } else if !alloc.provenance.ptrs.is_empty() {
+            // Codegen the provenance pointer.
+            trace!("codegen_const_ptr with_prov");
+            let ptr = alloc.provenance.ptrs[0];
+            let alloc_id = ptr.1.0;
+            let typ = self.codegen_ty_stable(ty);
+            self.codegen_alloc_pointer(typ, alloc_id, ptr.0, span)
         } else {
-            Expr::struct_expr_from_values(
-                overall_t,
-                vec![self.codegen_scalar(s, fty, span)],
-                &self.symbol_table,
-            )
+            // If there's no provenance, just codegen the pointer address.
+            trace!("codegen_const_ptr no_prov");
+            let expr = if self.symbol_table.machine_model().pointer_width == 32 {
+                Expr::int_constant(
+                    alloc.read_uint().unwrap() as u32,
+                    Type::unsigned_int(self.symbol_table.machine_model().pointer_width),
```

Comment:
Might want to replace `self.symbol_table.machine_model().pointer_width` with 32.